## Task 3

## Run the Decision Tree classifier on Image Segmentation data, Large Soybean Database, and Thyroid disease records with a 72% and 28% split of training and test data respectively. 

### Image Segmentation data
#### EDA 
#### Attributes 19
#### Classes 7
#### Data floats,integers

In [289]:
from scipy.io import arff
import pandas as pd

data = arff.loadarff('data/segment.arff')
df = pd.DataFrame(data[0])

df.head()

region-centroid-col  region-centroid-row  region-pixel-count  \
0                218.0                178.0                 9.0   
1                113.0                130.0                 9.0   
2                202.0                 41.0                 9.0   
3                 32.0                173.0                 9.0   
4                 61.0                197.0                 9.0   

   short-line-density-5  short-line-density-2  vedge-mean  vegde-sd  \
0              0.111111                   0.0    0.833333  0.547722   
1              0.000000                   0.0    0.277778  0.250924   
2              0.000000                   0.0    0.944448  0.772202   
3              0.000000                   0.0    1.722220  1.781590   
4              0.000000                   0.0    1.444440  1.515350   

   hedge-mean  hedge-sd  intensity-mean  rawred-mean  rawblue-mean  \
0    1.111110  0.544331       59.629600      52.4444      75.22220   
1    0.333333  0.365148        0.888889       0.0000       2.55556   
2    1.111110  1.025600      123.037000     111.8890     139.77800   
3    9.000000  6.749490       43.592600      39.5556      52.88890   
4    2.611110  1.925460       49.592600      44.2222      61.55560   

   rawgreen-mean  exred-mean  exblue-mean  exgreen-mean  value-mean  \
0      51.222200   -21.55560      46.7778     -25.22220    75.22220   
1       0.111111    -2.66667       5.0000      -2.33333     2.55556   
2     117.444000   -33.44440      50.2222     -16.77780   139.77800   
3      38.333300   -12.11110      27.8889     -15.77780    52.88890   
4      43.000000   -16.11110      35.8889     -19.77780    61.55560   

   saturation-mean  hue-mean    class  
0         0.318996  -2.04055     path  
1         1.000000  -2.12325  foliage  
2         0.199347  -2.29992      sky  
3         0.266914  -1.99886     path  
4         0.302925  -2.02227     path

In [290]:
from datacleaner import autoclean

my_clean_data = autoclean(df)

print(my_clean_data)

      region-centroid-col  region-centroid-row  region-pixel-count  \
0                   218.0                178.0                 9.0   
1                   113.0                130.0                 9.0   
2                   202.0                 41.0                 9.0   
3                    32.0                173.0                 9.0   
4                    61.0                197.0                 9.0   
5                   149.0                185.0                 9.0   
6                   197.0                229.0                 9.0   
7                    29.0                111.0                 9.0   
8                     1.0                 81.0                 9.0   
9                    69.0                 85.0                 9.0   
10                  152.0                 83.0                 9.0   
11                  248.0                153.0                 9.0   
12                  137.0                141.0                 9.0   
13                  

In [291]:
x = df[df.columns.drop('class')]
print(x)

      region-centroid-col  region-centroid-row  region-pixel-count  \
0                   218.0                178.0                 9.0   
1                   113.0                130.0                 9.0   
2                   202.0                 41.0                 9.0   
3                    32.0                173.0                 9.0   
4                    61.0                197.0                 9.0   
5                   149.0                185.0                 9.0   
6                   197.0                229.0                 9.0   
7                    29.0                111.0                 9.0   
8                     1.0                 81.0                 9.0   
9                    69.0                 85.0                 9.0   
10                  152.0                 83.0                 9.0   
11                  248.0                153.0                 9.0   
12                  137.0                141.0                 9.0   
13                  

In [292]:
y = df['class']
print(y)

0       4
1       2
2       5
3       4
4       4
5       4
6       3
7       0
8       2
9       0
10      1
11      6
12      6
13      4
14      3
15      6
16      0
17      4
18      1
19      1
20      6
21      0
22      2
23      5
24      3
25      6
26      2
27      4
28      5
29      2
       ..
2280    6
2281    6
2282    4
2283    6
2284    4
2285    1
2286    0
2287    4
2288    1
2289    2
2290    5
2291    0
2292    4
2293    1
2294    0
2295    5
2296    2
2297    2
2298    5
2299    2
2300    6
2301    6
2302    4
2303    4
2304    1
2305    0
2306    5
2307    1
2308    2
2309    0
Name: class, Length: 2310, dtype: int64


In [293]:
# import Decision tree classifier
from sklearn.tree import DecisionTreeClassifier
from sklearn import tree

clf = tree.DecisionTreeClassifier(random_state=0)#random_state will give consistent results
clf = clf.fit(x,y)

In [294]:
#number of nodes
treeObj = clf.tree_
print treeObj.node_count

143


In [295]:
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import confusion_matrix

In [296]:
Y_pred = cross_val_predict(clf,x,y,cv=10)
print(Y_pred)
conf_mat = confusion_matrix(y,Y_pred)
print (conf_mat)

[4 2 5 ..., 1 2 6]
[[323   0   5   0   0   0   2]
 [  2 311   8   0   0   0   9]
 [  1   0 309   0   1   0  19]
 [  0   1   0 329   0   0   0]
 [  0   0   0   1 329   0   0]
 [  0   0   0   0   0 330   0]
 [  0  10  25   0   0   0 295]]


In [297]:
#accuracy
from sklearn import metrics
print metrics.accuracy_score(y,Y_pred)

0.963636363636


In [298]:
#error rate

e = metrics.accuracy_score(y,Y_pred)
print 1-e

0.0363636363636


### Large Soybean Database

#### EDA 
#### Attributes 35
#### Classes 19
#### Data strings,integer range,missing data

In [299]:
from scipy.io import arff
import pandas as pd

data = arff.loadarff('data/soybean.arff')
df = pd.DataFrame(data[0])

df.head()

date plant-stand   precip  temp hail         crop-hist area-damaged  \
0  october      normal  gt-norm  norm  yes       same-lst-yr    low-areas   
1   august      normal  gt-norm  norm  yes  same-lst-two-yrs    scattered   
2     july      normal  gt-norm  norm  yes       same-lst-yr    scattered   
3     july      normal  gt-norm  norm  yes       same-lst-yr    scattered   
4  october      normal  gt-norm  norm  yes  same-lst-two-yrs    scattered   

     severity   seed-tmt germination          ...           sclerotia  \
0  pot-severe       none      90-100          ...              absent   
1      severe  fungicide       80-89          ...              absent   
2      severe  fungicide       lt-80          ...              absent   
3      severe       none       80-89          ...              absent   
4  pot-severe       none       lt-80          ...              absent   

  fruit-pods fruit-spots  seed mold-growth seed-discolor seed-size shriveling  \
0       norm         dna  norm      absent        absent      norm     absent   
1       norm         dna  norm      absent        absent      norm     absent   
2       norm         dna  norm      absent        absent      norm     absent   
3       norm         dna  norm      absent        absent      norm     absent   
4       norm         dna  norm      absent        absent      norm     absent   

  roots                  class  
0  norm  diaporthe-stem-canker  
1  norm  diaporthe-stem-canker  
2  norm  diaporthe-stem-canker  
3  norm  diaporthe-stem-canker  
4  norm  diaporthe-stem-canker  

[5 rows x 36 columns]

In [300]:
from datacleaner import autoclean

my_clean_data = autoclean(df)

print(my_clean_data)

     date  plant-stand  precip  temp  hail  crop-hist  area-damaged  severity  \
0       6            2       1     3     2          4             1         2   
1       2            2       1     3     2          3             2         3   
2       3            2       1     3     2          4             2         3   
3       3            2       1     3     2          4             2         3   
4       6            2       1     3     2          3             2         2   
5       7            2       1     3     2          2             2         2   
6       7            2       1     3     2          3             2         2   
7       2            2       1     3     1          4             2         2   
8       6            2       1     3     2          2             2         2   
9       2            2       1     3     2          3             2         3   
10      6            2       2     1     2          4             4         2   
11      2            2      

In [301]:
x = df[df.columns.drop('class')]
print(x)

     date  plant-stand  precip  temp  hail  crop-hist  area-damaged  severity  \
0       6            2       1     3     2          4             1         2   
1       2            2       1     3     2          3             2         3   
2       3            2       1     3     2          4             2         3   
3       3            2       1     3     2          4             2         3   
4       6            2       1     3     2          3             2         2   
5       7            2       1     3     2          2             2         2   
6       7            2       1     3     2          3             2         2   
7       2            2       1     3     1          4             2         2   
8       6            2       1     3     2          2             2         2   
9       2            2       1     3     2          3             2         3   
10      6            2       2     1     2          4             4         2   
11      2            2      

In [302]:
y = df['class']
print(y)

0      10
1      10
2      10
3      10
4      10
5      10
6      10
7      10
8      10
9      10
10      7
11      7
12      7
13      7
14      7
15      7
16      7
17      7
18      7
19      7
20     18
21     18
22     18
23     18
24     18
25     18
26     18
27     18
28     18
29     18
       ..
653     9
654     9
655     9
656     8
657     8
658     8
659     8
660     8
661     8
662     8
663     8
664     0
665     0
666     0
667     0
668     0
669     0
670     0
671     0
672     0
673     0
674     0
675     0
676     0
677     0
678     0
679    13
680    13
681    13
682    13
Name: class, Length: 683, dtype: int64


In [303]:
# import Decision tree classifier
from sklearn.tree import DecisionTreeClassifier
clf = DecisionTreeClassifier(random_state=0)

In [304]:
clf = clf.fit(x,y)

In [305]:
#number of nodes
treeObj = clf.tree_
print treeObj.node_count

151


In [306]:
Y_pred = cross_val_predict(clf,x,y,cv=10)
#print(Y_pred)
conf_mat = confusion_matrix(y,Y_pred)
print (conf_mat)

[[16  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0 72  0  0  0  1  0  0  0  0  0  0 18  0  0  0  0  0  0]
 [ 0  0 43  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1]
 [ 0  1  0 19  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0 19  0  0  0  0  0  0  0  0  0  0  0  0  1  0]
 [ 0  1  0  0  0 81  0  0  0  0  0  0  7  0  3  0  0  0  0]
 [ 0  0  0  0  0  0 44  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0 20  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0 14  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0 15  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0 20  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0 20  0  0  0  0  0  0  0]
 [ 0 11  0  0  0  8  1  0  0  0  0  0 71  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0  8  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0  0 20  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0 88  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0

In [307]:
#accuracy

print metrics.accuracy_score(y,Y_pred)

0.92093704246


In [308]:
e = metrics.accuracy_score(y,Y_pred)
print 1-e

0.0790629575403


### Thyroid disease records

#### EDA

#### Attributes 30

#### Classes 2

#### Data floats,integers,strings,characters,missing values

In [315]:
from scipy.io import arff
import pandas as pd

data = arff.loadarff('data/sick.arff')
df = pd.DataFrame(data[0])

df.head()

age sex on thyroxine query on thyroxine on antithyroid medication sick  \
0  41.0   F            f                  f                         f    f   
1  23.0   F            f                  f                         f    f   
2  46.0   M            f                  f                         f    f   
3  70.0   F            t                  f                         f    f   
4  70.0   F            f                  f                         f    f   

  pregnant thyroid surgery I131 treatment query hypothyroid    ...     \
0        f               f              f                 f    ...      
1        f               f              f                 f    ...      
2        f               f              f                 f    ...      
3        f               f              f                 f    ...      
4        f               f              f                 f    ...      

  TT4 measured    TT4 T4U measured   T4U FTI measured    FTI TBG measured  \
0            t  125.0            t  1.14            t  109.0            f   
1            t  102.0            f   NaN            f    NaN            f   
2            t  109.0            t  0.91            t  120.0            f   
3            t  175.0            f   NaN            f    NaN            f   
4            t   61.0            t  0.87            t   70.0            f   

   TBG referral source     Class  
0  NaN            SVHC  negative  
1  NaN           other  negative  
2  NaN           other  negative  
3  NaN           other  negative  
4  NaN             SVI  negative  

[5 rows x 30 columns]

In [316]:
from datacleaner import autoclean

my_clean_data = autoclean(df)

print(my_clean_data)


       age  sex  on thyroxine  query on thyroxine  on antithyroid medication  \
0     41.0    1             0                   0                          0   
1     23.0    1             0                   0                          0   
2     46.0    2             0                   0                          0   
3     70.0    1             1                   0                          0   
4     70.0    1             0                   0                          0   
5     18.0    1             1                   0                          0   
6     59.0    1             0                   0                          0   
7     80.0    1             0                   0                          0   
8     66.0    1             0                   0                          0   
9     68.0    2             0                   0                          0   
10    84.0    1             0                   0                          0   
11    67.0    1             1           

In [331]:
x = df[df.columns.drop('Class')]
#print(x)

df.TBG = df.TBG.fillna(value=0)

print(x)

       age  sex  on thyroxine  query on thyroxine  on antithyroid medication  \
0     41.0    1             0                   0                          0   
1     23.0    1             0                   0                          0   
2     46.0    2             0                   0                          0   
3     70.0    1             1                   0                          0   
4     70.0    1             0                   0                          0   
5     18.0    1             1                   0                          0   
6     59.0    1             0                   0                          0   
7     80.0    1             0                   0                          0   
8     66.0    1             0                   0                          0   
9     68.0    2             0                   0                          0   
10    84.0    1             0                   0                          0   
11    67.0    1             1           

In [332]:
y = df['Class']
print(y)

0       0
1       0
2       0
3       0
4       0
5       0
6       0
7       1
8       0
9       0
10      0
11      0
12      0
13      0
14      0
15      0
16      0
17      0
18      0
19      0
20      0
21      0
22      1
23      0
24      0
25      0
26      0
27      0
28      0
29      0
       ..
3742    0
3743    0
3744    0
3745    0
3746    0
3747    0
3748    0
3749    0
3750    1
3751    1
3752    0
3753    0
3754    0
3755    1
3756    0
3757    0
3758    0
3759    0
3760    0
3761    0
3762    0
3763    0
3764    0
3765    0
3766    0
3767    0
3768    0
3769    0
3770    0
3771    0
Name: Class, Length: 3772, dtype: int64


In [333]:
clf = clf.fit(x,y)

In [334]:
#number of nodes
treeObj = clf.tree_
print treeObj.node_count

83


In [335]:
Y_pred = cross_val_predict(clf,x,y,cv=10)
#print(Y_pred)
conf_mat = confusion_matrix(y,Y_pred)
print (conf_mat)

[[3521   20]
 [  26  205]]


In [336]:
#accuracy
print metrics.accuracy_score(y,Y_pred)

0.987804878049


In [337]:
#error rate

e = metrics.accuracy_score(y,Y_pred)
print 1-e

0.0121951219512


### Problem 5:

### Calculate and report the error-rate and accuracy for each of the datasets. Which of the 3 datasets has the highest number of correctly classified instances?




#### Dataset  Accuracy
#### Image Segmentation data :0.963636363636


#### Large Soybean Database : 0.92093704246


#### Thyroid disease records  : 0.987804878049


####  Thyroid disease records has highest accuracy and correctly classified instances.

##  Which of the three datasets has the smallest and largest decision trees? Explain why you think the size of the decision trees varies.

#### soybean dataset has 151 nodes which is the lagest tree.Sick dataset has 83 nodes which is the smallest tree.
#### Soybean dataset has more attributes and hence more conditions to form as a tree.
#### number of attributes ,classes and total number of data all participate in complexilty of tree.